packages/shared/pkg/reverse-proxy/proxy.go

```diff
@@ -0,0 +1,230 @@
+package reverse_proxy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"go.opentelemetry.io/otel/metric"
+	"go.uber.org/zap"
+
+	template "github.com/e2b-dev/infra/packages/shared/pkg/reverse-proxy/error-template"
+)
+
+const (
+	maxConnectionDuration = 24 * time.Hour // The same as the current max sandbox duration.
+	maxIdleConnections    = 32768          // Reasonably big number that is lower than the number of available ports.
+)
+
+type routingTargetContextKey struct{}
+
+// RoutingTarget contains information about where to route the request.
+type RoutingTarget struct {
+	Url       *url.URL
+	SandboxId string
+	Logger    *zap.Logger
+	// ConnectionKey is used for identifying which keepalive connections are not the same so we can prevent uninted reuse.
+	// This is evaluated before checking for existing connection to the IP:port pair.
+	ConnectionKey string
+}
+
+type Proxy struct {
+	http.Server
+	totalEstablishedConnections *atomic.Uint64
+}
+
+func (p *Proxy) TotalEstablishedConnections() uint64 {
+	return p.totalEstablishedConnections.Load()
+}
+
+func New(
+	port uint,
+	idleTimeout time.Duration,
+	connectionTimeout time.Duration,
+	activeConnections *metric.Int64UpDownCounter,
+	getRoutingTarget func(r *http.Request) (*RoutingTarget, error),
+) *Proxy {
+	handler, counter := proxyHandler(
+		getRoutingTarget,
+		activeConnections,
+		idleTimeout,
+		connectionTimeout,
+	)
+
+	return &Proxy{
+		Server: http.Server{
+			Addr:              fmt.Sprintf(":%d", port),
+			ReadTimeout:       maxConnectionDuration,
+			WriteTimeout:      maxConnectionDuration,
+			IdleTimeout:       idleTimeout,
+			ReadHeaderTimeout: 20 * time.Second,
+			Handler:           http.HandlerFunc(handler),
+		},
+		totalEstablishedConnections: counter,
+	}
+}
+
+func proxyHandler(
+	getRoutingTarget func(r *http.Request) (*RoutingTarget, error),
+	activeConnections *metric.Int64UpDownCounter,
+	idleTimeout time.Duration,
+	connectionTimeout time.Duration,
+) (func(w http.ResponseWriter, r *http.Request), *atomic.Uint64) {
+	var counter atomic.Uint64
+
+	var proxiesMu sync.Mutex
+
+	proxies := expirable.NewLRU[string, *httputil.ReverseProxy](0, nil, maxConnectionDuration)
+
+	getProxy := func(connectionKey string) *httputil.ReverseProxy {
+		proxiesMu.Lock()
+		defer proxiesMu.Unlock()
+
+		proxy, ok := proxies.Get(connectionKey)
+		if ok {
+			return proxy
+		}
+
+		transport := &http.Transport{
```

Comment:
By creating more transports this way, RAM consumption will go up a lot. In the initial setup, we didn't reuse the transport. 

Reusing them for each sandbox should work well in the orchestrator, but it might be an issue in the client proxy. Could you check the impact on memory before and after this change by simulating increased traffic?